src/backend/access/transam/timeline.c

```diff
@@ -174,10 +175,27 @@ readTimeLineHistory(TimeLineID targetTLI)
 
 		lasttli = tli;
 
+		end = ((uint64) (switchpoint_hi)) << 32 | (uint64) switchpoint_lo;
+
+		/*
+		 * If start pointer of newest timeline is less than previous, we should
+		 * ignore all previous timeline. It may be when recovery was done from
+		 * old archive (two or more timelines before) with promote. At this case
+		 * previous timeline should be ignored, because the current timeline was
+		 * promoted from another timeline (not previous).
+		 */
+		if (entry && prevend > end)
+		{
```

Comment:
how about
```suggestion
		if (prevend > end)
		{
			Assert(entry != NULL);
```
and remove `= NULL` change from function begin?

Reply:
Done.

Reply:
I reverted the last changes because of compilation error:
```
#4 78.66 timeline.c: In function ‘readTimeLineHistory’:
#4 78.66 timeline.c:192:36: error: ‘entry’ may be used uninitialized in this function [-Werror=maybe-uninitialized]
#4 78.66   192 |                         entry->end = end;
#4 78.66       |                         ~~~~~~~~~~~^~~~~
#4 78.81 cc1: all warnings being treated as errors
#4 78.81 make[5]: *** [../../../../src/Makefile.global:1026: timeline.o] Error 1
#4 78.81 make[5]: *** Waiting for unfinished jobs....
#4 79.75 make[4]: *** [../../../src/backend/common.mk:39: transam-recursive] Error 2
#4 79.75 make[4]: *** Waiting for unfinished jobs....
#4 80.73 fe-secure-openssl.c:1042:19: warning: 'ENGINE_by_id' is deprecated [-Wdeprecat
```

Reply:
hm...